src/realm.h

```diff
@@ -1559,37 +1559,271 @@ RLM_API void realm_collection_changes_get_ranges(
 
+/**
+ * Create a `realm_dictionary_t` from a pointer to a `realm::object_store::Dictionary`,
+ * copy-constructing the internal representation.
+ *
+ * @param pdict A pointer to an instance of `realm::object_store::Dictionary`.
+ * @param n Must be equal to `sizeof(realm::object_store::Dictionary)`.
+ * @return A non-null pointer if no exception occurred.
+ */
 RLM_API realm_dictionary_t* _realm_dictionary_from_native_copy(const void* pdict, size_t n);
+
+/**
+ * Create a `realm_dictionary_t` from a pointer to a `realm::object_store::Dictionary`,
+ * move-constructing the internal representation.
+ *
+ * @param pdict A pointer to an instance of `realm::object_store::Dictionary`.
+ * @param n Must be equal to `sizeof(realm::object_store::Dictionary)`.
+ * @return A non-null pointer if no exception occurred.
+ */
 RLM_API realm_dictionary_t* _realm_dictionary_from_native_move(void* pdict, size_t n);
-RLM_API realm_dictionary_t* realm_get_dictionary(const realm_object_t*, realm_property_key_t);
-RLM_API size_t realm_dictionary_size(const realm_dictionary_t*);
-RLM_API bool realm_dictionary_get(const realm_dictionary_t*, realm_value_t key, realm_value_t* out_value,
-                                  bool* out_found);
-RLM_API bool realm_dictionary_insert(realm_dictionary_t*, realm_value_t key, realm_value_t value, bool* out_inserted,
-                                     size_t* out_index);
+
+/**
+ * Get the size of a dictionary (the number of unique keys).
+ *
+ * This function may fail if the object owning the dictionary has been deleted.
+ *
+ * @param out_size Where to put the dictionary size. May be NULL.
+ * @return True if no exception occurred.
+ */
+RLM_API bool realm_dictionary_size(const realm_dictionary_t*, size_t* out_size);
+
+
+/**
+ * Get the property that this dictionary came from.
+ *
+ * @return True if no exception occurred.
+ */
+RLM_API bool realm_dictionary_get_property(const realm_dictionary_t*, realm_property_info_t* out_info);
+
+/**
+ * Find an element in a dictionary.
+ *
+ * @param key The key to look for.
+ * @param out_value If non-null, the value for the corresponding key.
+ * @param out_found If non-null, will be set to true if the dictionary contained the key.
+ * @return True if no exception occurred.
+ */
+RLM_API bool realm_dictionary_find(const realm_dictionary_t*, realm_value_t key, realm_value_t* out_value,
+                                   bool* out_found);
+
+/**
+ * Get the key-value pair at @a index.
+ *
+ * Note that the indices of elements in the dictionary move around as other
+ * elements are inserted/removed.
+ *
+ * @param index The index in the dictionary.
+ * @param out_key If non-null, will be set to the key at the corresponding index.
+ * @param out_value If non-null, will be set to the value at the corresponding index.
+ * @return True if no exception occurred.
+ */
+RLM_API bool realm_dictionary_get(const realm_dictionary_t*, size_t index, realm_value_t* out_key,
```

Comment:
In Java/Kotlin we have operations to get `keys()` or `values()`. Don't know if it is worth having specific functions for these cases too?